api/providers/tasks.py

```diff
@@ -0,0 +1,539 @@
+                fullname=initiator.fullname,
+                auto_approval=auto_approval,
+                total=initial_row_count,
+                successes=successful_row_count,
+                draft_errors=draft_error_list,
+                failures=len(draft_error_list),
+                pending_submissions_url=get_provider_submission_url(provider),
+            )
+        elif upload.state == JobState.DONE_ERROR:
+            mails.send_mail(
+                to_addr=initiator.username,
+                mail=mails.REGISTRATION_BULK_UPLOAD_FAILURE_ALL,
+                fullname=initiator.fullname,
+                count=initial_row_count,
+                draft_errors=draft_error_list,
+            )
+        else:
+            message = 'Failed to send registration bulk upload outcome email due to invalid ' \
+                      'upload state: [upload={}, state={}]'.format(upload.id, upload.state.name)
+            logger.error(message)
+            sentry.log_message(message)
+        logger.info('Email sent to bulk upload initiator [{}]'.format(initiator._id))
+
+
+def handle_registration_row(row, initiator, provider, schema, auto_approval=False, dry_run=True):
+    """Create a draft registration for one registration row in a given bulk upload job.
+    """
+
+    metadata = row.csv_parsed.get('metadata', {}) or {}
+    row_external_id = metadata.get('External ID', 'N/A')
+    row_title = metadata.get('Title', 'N/A')
+    responses = row.csv_parsed.get('registration_responses', {}) or {}
+    auth = Auth(user=initiator)
+
+    # Check node
+    node = None
+    node_id = metadata.get('Project GUID')
+    if node_id:
+        try:
+            node = AbstractNode.objects.get(guids___id=node_id, is_deleted=False, type='osf.node')
+            initiator_contributor = node.contributor_set.get(user=initiator)
+        except AbstractNode.DoesNotExist:
+            error = 'Node does not exist: [node_id={}]'.format(node_id)
+            raise RegistrationBulkCreationRowError(row.upload.id, row.id, row_title, row_external_id, error=error)
+        except AbstractNode.MultipleObjectsReturned:
+            error = 'Multiple nodes returned: [node_id={}]'.format(node_id)
+            raise RegistrationBulkCreationRowError(row.upload.id, row.id, row_title, row_external_id, error=error)
+        except Contributor.DoesNotExist:
+            error = 'Initiator [{}] must be a contributor on the project [{}]'.format(initiator._id, node._id)
+            raise RegistrationBulkCreationRowError(row.upload.id, row.id, row_title, row_external_id, error=error)
+        if initiator_contributor.permission not in [WRITE, ADMIN]:
+            error = 'Initiator [{}] must at least have WRITE permission on the project [{}]'.format(initiator._id, node._id)
+            raise RegistrationBulkCreationRowError(row.upload.id, row.id, row_title, row_external_id, error=error)
+
+    # Prepare subjects
+    subject_texts = metadata.get('Subjects', []) or []
```

Comment:
The parser is not perfect. For some fields, it returns empty string instead of empty `dict` or `list`. Thus the task has to add this extra step everywhere until we have a more robust parser. This is on the post-release coding list.